x-pack/plugins/cloud/public/plugin.tsx

```diff
@@ -230,109 +242,158 @@ export class CloudPlugin implements Plugin<CloudSetup> {
-                    // https://help.fullstory.com/hc/en-us/articles/1500004101581-FS-setVars-API-Sending-custom-page-data-to-FullStory
-                    pageName: `${compact([name, page]).join(':')}`,
-                    app_id_str: name ?? 'unknown',
-                    page_str: page,
-                    ent_id_str: id,
-                  },
-                  isUndefined
-                )
-              );
-            });
+    // This needs to be called synchronously to be sure that we populate the user ID soon enough to make sessions merging
+    // across domains work
+    if (security) {
+      analytics.registerContextProvider({
+        name: 'cloud_user_id',
+        context$: from(loadUserId({ getCurrentUser: security.authc.getCurrentUser })).pipe(
+          filter((userId): userId is string => Boolean(userId)),
+          exhaustMap(async (userId) => {
+            const { sha256 } = await import('js-sha256');
+            // Join the cloud org id and the user to create a truly unique user id.
+            // The hashing here is to keep it at clear as possible in our source code that we do not send literal user IDs
+            return { userId: sha256(esOrgId ? `${esOrgId}:${userId}` : `${userId}`) };
           })
-          .catch((e) => {
-            // eslint-disable-next-line no-console
-            console.error(
-              `[cloud.full_story] Could not retrieve application service due to error: ${e.toString()}`,
-              e
-            );
-          });
-        const kibanaVer = this.initializerContext.env.packageInfo.version;
-        // TODO: use semver instead
-        const parsedVer = (kibanaVer.indexOf('.') > -1 ? kibanaVer.split('.') : []).map((s) =>
-          parseInt(s, 10)
-        );
-        // `str` suffix is required for evn vars, see docs: https://help.fullstory.com/hc/en-us/articles/360020623234
-        fullStory.identify(hashedId, {
-          version_str: kibanaVer,
-          version_major_int: parsedVer[0] ?? -1,
-          version_minor_int: parsedVer[1] ?? -1,
-          version_patch_int: parsedVer[2] ?? -1,
-          org_id_str: esOrgId,
-        });
-      }
-    } catch (e) {
-      // eslint-disable-next-line no-console
-      console.error(
-        `[cloud.full_story] Could not call FS.identify due to error: ${e.toString()}`,
-        e
-      );
+        ),
+        schema: {
+          userId: {
+            type: 'keyword',
+            _meta: { description: 'The user id scoped as seen by Cloud (hashed)' },
+          },
+        },
+      });
     }
 
+    const executionContext = await executionContextPromise;
+    analytics.registerContextProvider({
+      name: 'execution_context',
+      context$: executionContext.context$.pipe(
+        // Update the current context every time it changes
+        map(({ name, page, id }) => ({
+          pageName: `${compact([name, page]).join(':')}`,
+          app_id: name ?? 'unknown',
```

Comment:
If it make sense to have the `cloud` plugin register the orgId/userId providers, I wonder if that's the best owner for the EC and version providers?

Reply:
I totally agree! I just used this PR to "convert" the FS calls to the new analytics client's API. The plan is to use https://github.com/elastic/kibana/issues/125690 to move them where they actually belong :)